libs/types/src/cdf/ballot-definition/convert.ts

```diff
@@ -0,0 +1,412 @@
+              {
+                '@type': 'BallotDefinition.ExternalIdentifier',
+                Type: Cdf.IdentifierType.StateLevel,
+                Value: ballotStyle.id,
+              },
+            ],
+          })
+        ),
+      },
+    ],
+
+    Party: vxfElection.parties.map((party) => ({
+      '@type': 'BallotDefinition.Party',
+      '@id': party.id,
+      Name: text(party.fullName),
+      Abbreviation: text(party.abbrev),
+      vxBallotLabel: text(party.name),
+    })),
+
+    GpUnit: [
+      {
+        '@type': 'BallotDefinition.ReportingUnit',
+        '@id': stateId,
+        Name: text(vxfElection.state),
+        Type: Cdf.ReportingUnitType.State,
+        ComposingGpUnitIds: [vxfElection.county.id],
+      },
+      {
+        '@type': 'BallotDefinition.ReportingUnit',
+        '@id': vxfElection.county.id,
+        Name: text(vxfElection.county.name),
+        Type: Cdf.ReportingUnitType.County,
+        ComposingGpUnitIds: vxfElection.districts.map(
+          (district) => district.id
+        ),
+      },
+      ...vxfElection.districts.map(
+        (district): Cdf.ReportingUnit => ({
+          '@type': 'BallotDefinition.ReportingUnit',
+          '@id': district.id,
+          Name: text(district.name),
+          // Since we represent multiple real-world entities as districts in VXF,
+          // we can't know the actual type to use here
+          Type: Cdf.ReportingUnitType.Other,
+          // To figure out which precincts are in this district, we look at the
+          // associated ballot styles
+          ComposingGpUnitIds: unique(
+            vxfElection.ballotStyles
+              .filter((ballotStyle) =>
+                ballotStyle.districts.includes(district.id)
+              )
+              .flatMap((ballotStyle) => ballotStyle.precincts)
+          ),
+        })
+      ),
+      ...vxfElection.precincts.map(
+        (precinct): Cdf.ReportingUnit => ({
+          '@type': 'BallotDefinition.ReportingUnit',
+          '@id': precinct.id,
+          Name: text(precinct.name),
+          Type: Cdf.ReportingUnitType.Precinct,
+        })
+      ),
+    ],
+
+    GeneratedDate: dateTimeString(new Date()),
+    Issuer: 'VotingWorks',
+    IssuerAbbreviation: 'VX',
+    VendorApplicationId: 'VxSuite',
+    Version: Cdf.BallotDefinitionVersion.v1_0_0,
+    SequenceStart: 1,
+    SequenceEnd: 1,
+  };
+}
+
+// TODO Should we make a tighter BallotDefinition type that encodes some of our constraints?
+export function convertCdfBallotDefinitionToVxfElection(
+  cdfBallotDefinition: Cdf.BallotDefinition
+): Vxf.Election {
+  assert(cdfBallotDefinition.Election);
+  const election = assertDefined(cdfBallotDefinition.Election[0]);
+  const gpUnits = assertDefined(cdfBallotDefinition.GpUnit);
+
+  const state = find(
+    gpUnits,
+    (gpUnit) => gpUnit.Type === Cdf.ReportingUnitType.State
+  );
+  const county = find(
+    gpUnits,
+    (gpUnit) => gpUnit.Type === Cdf.ReportingUnitType.County
+  );
```

Comment:
Are these guaranteed to exist? Is that one of the constraints you'd like to enforce?

Reply:
Hmm, good question. I think I put this logic in when I was only thinking about round trip conversion from VXF<->CDF. I wonder how many places in the product actual use `state` and `county`. Assuming there are some, I would imagine it makes sense to enforce this as a constraint to start.